src/verify.rs

```diff
@@ -0,0 +1,612 @@
+    ChainedBeaconNeedsPreviousSignature,
+    #[error("invalid signature length")]
+    InvalidSignatureLength,
+    #[error("invalid public key")]
+    InvalidPublicKey,
+    #[error("message can't be empty")]
+    EmptyMessage,
+    #[error("signature verification failed")]
+    SignatureFailedVerification,
+    #[error("the randomness for the beacon did not match the signature")]
+    InvalidRandomness,
+}
+
+pub fn verify_beacon(
+    scheme_id: &SchemeID,
+    public_key: &[u8],
+    beacon: &Beacon,
+) -> Result<(), VerificationError> {
+    if Sha256::digest(&beacon.signature).to_vec() != beacon.randomness {
+        return Err(VerificationError::InvalidRandomness);
+    }
+    match scheme_id {
+        SchemeID::PedersenBlsChained => verify_on_g2(
+            public_key,
+            &chained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::PedersenBlsUnchained => verify_on_g2(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::UnchainedOnG1 => verify_on_g1(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::UnchainedOnG1RFC9380 => verify_on_g1(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G1,
+        ),
+    }
+}
+
+fn unchained_beacon_message(beacon: &Beacon) -> Result<Vec<u8>, VerificationError> {
+    if !beacon.previous_signature.is_empty() {
+        return Err(VerificationError::UnchainedHasPreviousSignature);
+    }
+    let round_bytes = beacon.round_number.to_be_bytes();
+
+    Ok(Sha256::digest(&round_bytes).to_vec())
+}
+
+fn chained_beacon_message(beacon: &Beacon) -> Result<Vec<u8>, VerificationError> {
+    if beacon.previous_signature.is_empty() {
+        Err(VerificationError::ChainedBeaconNeedsPreviousSignature)
+    } else {
+        // surely there's a better way to concat two slices
+        let mut message = Vec::new();
+        message.extend_from_slice(beacon.previous_signature.as_slice());
+        message.extend_from_slice(&beacon.round_number.to_be_bytes());
+        Ok(Sha256::digest(message.as_slice()).to_vec())
+    }
+}
+
+pub fn verify_on_g2(
+    public_key: &[u8],
+    message: &[u8],
+    signature: &[u8],
+    domain_separation_tag: &str,
+) -> Result<(), VerificationError> {
+    let pub_key_bytes: &[u8; 48] = public_key
+        .try_into()
+        .map_err(|_| VerificationError::InvalidPublicKey)?;
+
+    let sig_bytes: &[u8; 96] = signature
+        .try_into()
+        .map_err(|_| VerificationError::InvalidSignatureLength)?;
+
+    let p = G1Affine::from_compressed(pub_key_bytes).unwrap_or(G1Affine::identity());
+
+    let q = G2Affine::from_compressed(sig_bytes).unwrap_or(G2Affine::identity());
```

Comment:
Why the `unwrap_or`? Shouldn't that be an error case?
Feels like a case where `?` is useful.

Reply:
The ZkCrypto lib returns their own `CtOption` type that wraps a `u8` instead of a `Result` >.> makes sense perhaps if you want constant time ops, but we don't care.
I could perhaps write something to turn `CtOption` into a `Result` and give error messages; as-is, the infinity point will fail at later steps, but perhaps that's error-prone